### DISCUSSION TITLE
Can we make new query method, whenWhere?

### DISCUSSION THREAD

**Discussion by yogeshgalav** [laravel/framework#49190] [Ideas]
Sometimes I need to define two closures first for when and second for where,
and pass data in both,
it would be better if a single function would be available.

consider the following example.
```
->when($condition, function($query) use ($data) {
    $query->where(function($q) use ($data) {
        $q->where(...)->orWhere(...);
    });
})
```
the same code would be shorter and easy to read if it could be converted to following.
```
->whenWhere($condition, function($query) use ($data) {
        $query->where(...)->orWhere(...);
})
```

**Comment by henzeb** [comment]
Not sure I can understand the difference here... The second one simply seem to collapse the first, and could easily be done with when. I may be missing something here?

**Reply by yogeshgalav** [reply]
Yes, second one is the collapse or short representation of first one,
When clause does not act as super where condition,
or add brackets to query.
```
->whereNull('deleted_at')
->where(function($q) use ($data) {        
     $q->where('name', 'ILIKE', '%'.$request->search.'%')
         ->orWhere('name_kana', 'ILIKE', '%'.$request->search.'%');
});
 ```

**Comment by jbrooksuk** [comment]
Is this different to this?

```php
User::query()
  ->when(true, fn ($query) => $query->where('name', 'like', '%Taylor%')->orWhere('name', 'like', '%James%'))
  ->get();
```

**Reply by yogeshgalav** [reply]
Yes some condition like status or deleted_at you always want AND condition.
if conditions are in orWhere then it will break the query,
hence in sql it should look like
```
where 'status' =1  AND 
('name' LIKE '%Taylor%'  OR 'name' like '%James%')
```
For this you require external where closure irrespective of when closure.
